File: gematria/datasets/python/bhive_importer_test.py
# Copyright 2023 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from absl.testing import absltest

from gematria.datasets.python import bhive_importer
from gematria.llvm.python import canonicalizer
from gematria.llvm.python import llvm_architecture_support
from gematria.proto import basic_block_pb2
from gematria.proto import canonicalized_instruction_pb2
from gematria.proto import throughput_pb2

_CanonicalizedOperandProto = (
    canonicalized_instruction_pb2.CanonicalizedOperandProto)
_CanonicalizedInstructionProto = (
    canonicalized_instruction_pb2.CanonicalizedInstructionProto)

# A basic block that can be obtained by disassembling the basic block
# "4829d38b44246c8b54246848c1fb034829d04839c3" and using base_address=600.
_EXPECTED_BASIC_BLOCK_PROTO = basic_block_pb2.BasicBlockProto(
    machine_instructions=(
        basic_block_pb2.MachineInstructionProto(
            assembly="\tsubq\t%rdx, %rbx",
            machine_code=b"H)\323",
            address=600,
        ),
        basic_block_pb2.MachineInstructionProto(
            assembly="\tmovl\t108(%rsp), %eax",
            machine_code=b"\213D$l",
            address=603,
        ),
        basic_block_pb2.MachineInstructionProto(
            assembly="\tmovl\t104(%rsp), %edx",
            machine_code=b"\213T$h",
            address=607,
        ),
        basic_block_pb2.MachineInstructionProto(
            assembly="\tsarq\t$3, %rbx",
            machine_code=b"H\301\373\003",
            address=611,
        ),
        basic_block_pb2.MachineInstructionProto(
            assembly="\tsubq\t%rdx, %rax",
            machine_code=b"H)\320",
            address=615,
        ),
        basic_block_pb2.MachineInstructionProto(
            assembly="\tcmpq\t%rax, %rbx",
            machine_code=b"H9\303",
            address=618,
        ),
    ),
    canonicalized_instructions=(
        _CanonicalizedInstructionProto(
            mnemonic="SUB",
            llvm_mnemonic="SUB64rr",
            output_operands=(_CanonicalizedOperandProto(register_name="RBX"),),
            input_operands=(
                _CanonicalizedOperandProto(register_name="RBX"),
                _CanonicalizedOperandProto(register_name="RDX"),
            ),
            implicit_output_operands=(_CanonicalizedOperandProto(
                register_name="EFLAGS"),),
        ),
        _CanonicalizedInstructionProto(
            mnemonic="MOV",
            llvm_mnemonic="MOV32rm",
            output_operands=(_CanonicalizedOperandProto(register_name="EAX"),),
            input_operands=(
                _CanonicalizedOperandProto(
                    memory=_CanonicalizedOperandProto.MemoryLocation(
                        alias_group_id=1)),
                _CanonicalizedOperandProto(
                    address=_CanonicalizedOperandProto.AddressTuple(
                        base_register="RSP", displacement=108, scaling=1)),
            ),
        ),
        _CanonicalizedInstructionProto(
            mnemonic="MOV",
            llvm_mnemonic="MOV32rm",
            output_operands=(_CanonicalizedOperandProto(register_name="EDX"),),
            input_operands=(
                _CanonicalizedOperandProto(
                    memory=_CanonicalizedOperandProto.MemoryLocation(
                        alias_group_id=1)),
                _CanonicalizedOperandProto(
                    address=_CanonicalizedOperandProto.AddressTuple(
                        base_register="RSP", displacement=104, scaling=1)),
            ),
        ),
        _CanonicalizedInstructionProto(
            mnemonic="SAR",
            llvm_mnemonic="SAR64ri",
            output_operands=(_CanonicalizedOperandProto(register_name="RBX"),),
            input_operands=(
                _CanonicalizedOperandProto(register_name="RBX"),
                _CanonicalizedOperandProto(immediate_value=3),
            ),
            implicit_output_operands=(_CanonicalizedOperandProto(
                register_name="EFLAGS"),),
        ),
        _CanonicalizedInstructionProto(
            mnemonic="SUB",
            llvm_mnemonic="SUB64rr",
            output_operands=(_CanonicalizedOperandProto(register_name="RAX"),),
            input_operands=(
                _CanonicalizedOperandProto(register_name="RAX"),
                _CanonicalizedOperandProto(register_name="RDX"),
            ),
            implicit_output_operands=(_CanonicalizedOperandProto(
                register_name="EFLAGS"),),
        ),
        _CanonicalizedInstructionProto(
            mnemonic="CMP",
            llvm_mnemonic="CMP64rr",
            input_operands=(
                _CanonicalizedOperandProto(register_name="RBX"),
                _CanonicalizedOperandProto(register_name="RAX"),
            ),
            implicit_output_operands=(_CanonicalizedOperandProto(
                register_name="EFLAGS"),),
        ),
    ),
)


class BhiveImporterTest(absltest.TestCase):

  def setUp(self):
    super().setUp()

    self._x86_llvm = llvm_architecture_support.LlvmArchitectureSupport.x86_64()
    self._x86_canonicalizer = canonicalizer.Canonicalizer.x86_64(self._x86_llvm)

  def test_x86_basic_block_proto_from_bytes(self):
    importer = bhive_importer.BHiveImporter(self._x86_canonicalizer)
    block_proto = importer.basic_block_proto_from_bytes(
        b"\x90", base_address=999)
    self.assertEqual(
        block_proto,
        basic_block_pb2.BasicBlockProto(
            machine_instructions=(basic_block_pb2.MachineInstructionProto(
                assembly="\tnop",
                machine_code=b"\x90",
                address=999,
            ),),
            canonicalized_instructions=(_CanonicalizedInstructionProto(
                mnemonic="nop",
                llvm_mnemonic="NOOP",
            ),),
        ),
    )

  def test_x86_basic_block_proto_from_hex(self):
    importer = bhive_importer.BHiveImporter(self._x86_canonicalizer)
    # Basic block:
    #   sub rbx, rdx
    #   mov eax, dword ptr [rsp + 108]
    #   mov edx, dword ptr [rsp + 104]
    #   sar rbx, 3
    #   sub rax, rdx
    #   cmp rbx, rax
    block_proto = importer.basic_block_proto_from_hex(
        "4829d38b44246c8b54246848c1fb034829d04839c3",
        base_address=600,
    )
    self.assertEqual(block_proto, _EXPECTED_BASIC_BLOCK_PROTO)

  def test_x86_parse_csv_line(self):
    source_name = "test: made-up"
    importer = bhive_importer.BHiveImporter(self._x86_canonicalizer)
    # Same basic block as in test_x86_basic_block_proto_from_bytes().
    block_proto = importer.basic_block_with_throughput_proto_from_csv_line(
        source_name=source_name,
        line="4829d38b44246c8b54246848c1fb034829d04839c3,10",
        base_address=600,
        throughput_scaling=2.0,
    )
    self.assertEqual(
        block_proto,
        throughput_pb2.BasicBlockWithThroughputProto(
            basic_block=_EXPECTED_BASIC_BLOCK_PROTO,
            inverse_throughputs=(throughput_pb2.ThroughputWithSourceProto(
                source=source_name,
                inverse_throughput_cycles=[20.0],
            ),),
        ),
    )


if __name__ == "__main__":
  absltest.main()

File: gematria/datasets/bhive_importer_test.cc
// Copyright 2023 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "gematria/datasets/bhive_importer.h"

#include <memory>

#include "gematria/llvm/canonicalizer.h"
#include "gematria/llvm/llvm_architecture_support.h"
#include "gematria/testing/matchers.h"
#include "gmock/gmock.h"
#include "gtest/gtest.h"

namespace gematria {
namespace {

class BHiveImporterTest : public ::testing::Test {
 protected:
  static constexpr absl::string_view kLlvmTriple = "x86_64-unknown-unknown";
  static constexpr absl::string_view kSourceName = "bhive: skl";

  static constexpr double kScaling = 1.0 / 100.0;

  void SetUp() override {
    x86_llvm_ = LlvmArchitectureSupport::X86_64();
    x86_canonicalizer_ =
        std::make_unique<X86Canonicalizer>(&x86_llvm_->target_machine());
    x86_bhive_importer_ =
        std::make_unique<BHiveImporter>(x86_canonicalizer_.get());
  }

  std::unique_ptr<LlvmArchitectureSupport> x86_llvm_;
  std::unique_ptr<Canonicalizer> x86_canonicalizer_;
  std::unique_ptr<BHiveImporter> x86_bhive_importer_;
};

using BHiveImporterDeathTest = BHiveImporterTest;

TEST_F(BHiveImporterTest, EmptyBlock) {
  EXPECT_THAT(
      x86_bhive_importer_->ParseBHiveCsvLine(kSourceName, ",0", 0, 1, kScaling),
      IsOkAndHolds(EqualsProto(R"pb(basic_block {}
                                    inverse_throughputs {
                                      source: "bhive: skl"
                                      inverse_throughput_cycles: 0
                                    })pb")));
}

TEST_F(BHiveImporterTest, OneInstruction) {
  EXPECT_THAT(x86_bhive_importer_->ParseBHiveCsvLine(
                  kSourceName, "4929d2,100.000000", 0, 1, 0.5),
              IsOkAndHolds(EqualsProto(
                  R"pb(basic_block {
                         machine_instructions {
                           assembly: "\tsubq\t%rdx, %r10"
                           machine_code: "I)\322"
                         }
                         canonicalized_instructions {
                           mnemonic: "SUB"
                           llvm_mnemonic: "SUB64rr"
                           output_operands { register_name: "R10" }
                           input_operands { register_name: "R10" }
                           input_operands { register_name: "RDX" }
                           implicit_output_operands { register_name: "EFLAGS" }
                         }
                       }
                       inverse_throughputs {
                         source: "bhive: skl"
                         inverse_throughput_cycles: 50
                       })pb")));
  EXPECT_THAT(x86_bhive_importer_->BasicBlockProtoFromMachineCodeHex("4929d2"),
              IsOkAndHolds(EqualsProto(
                  R"pb(machine_instructions {
                         assembly: "\tsubq\t%rdx, %r10"
                         machine_code: "I)\322"
                       }
                       canonicalized_instructions {
                         mnemonic: "SUB"
                         llvm_mnemonic: "SUB64rr"
                         output_operands { register_name: "R10" }
                         input_operands { register_name: "R10" }
                         input_operands { register_name: "RDX" }
                         implicit_output_operands { register_name: "EFLAGS" }
                       })pb")));
}

TEST_F(BHiveImporterTest, MultipleInstructions) {
  static constexpr absl::string_view kExpectedBasicBlockProto =
      R"pb(basic_block {
             machine_instructions {
               assembly: "\tsubq\t%rdx, %rbx"
               machine_code: "H)\323"
               address: 100
             }
             machine_instructions {
               assembly: "\tmovl\t108(%rsp), %eax"
               machine_code: "\213D$l"
               address: 103
             }
             machine_instructions {
               assembly: "\tmovl\t104(%rsp), %edx"
               machine_code: "\213T$h"
               address: 107
             }
             machine_instructions {
               assembly: "\tsarq\t$3, %rbx"
               machine_code: "H\301\373\003"
               address: 111
             }
             machine_instructions {
               assembly: "\tsubq\t%rdx, %rax"
               machine_code: "H)\320"
               address: 115
             }
             machine_instructions {
               assembly: "\tcmpq\t%rax, %rbx"
               machine_code: "H9\303"
               address: 118
             }
             canonicalized_instructions {
               mnemonic: "SUB"
               llvm_mnemonic: "SUB64rr"
               output_operands { register_name: "RBX" }
               input_operands { register_name: "RBX" }
               input_operands { register_name: "RDX" }
               implicit_output_operands { register_name: "EFLAGS" }
             }
             canonicalized_instructions {
               mnemonic: "MOV"
               llvm_mnemonic: "MOV32rm"
               output_operands { register_name: "EAX" }
               input_operands { memory { alias_group_id: 1 } }
               input_operands {
                 address { base_register: "RSP" displacement: 108 scaling: 1 }
               }
             }
             canonicalized_instructions {
               mnemonic: "MOV"
               llvm_mnemonic: "MOV32rm"
               output_operands { register_name: "EDX" }
               input_operands { memory { alias_group_id: 1 } }
               input_operands {
                 address { base_register: "RSP" displacement: 104 scaling: 1 }
               }
             }
             canonicalized_instructions {
               mnemonic: "SAR"
               llvm_mnemonic: "SAR64ri"
               output_operands { register_name: "RBX" }
               input_operands { register_name: "RBX" }
               input_operands { immediate_value: 3 }
               implicit_output_operands { register_name: "EFLAGS" }
             }
             canonicalized_instructions {
               mnemonic: "SUB"
               llvm_mnemonic: "SUB64rr"
               output_operands { register_name: "RAX" }
               input_operands { register_name: "RAX" }
               input_operands { register_name: "RDX" }
               implicit_output_operands { register_name: "EFLAGS" }
             }
             canonicalized_instructions {
               mnemonic: "CMP"
               llvm_mnemonic: "CMP64rr"
               input_operands { register_name: "RBX" }
               input_operands { register_name: "RAX" }
               implicit_output_operands { register_name: "EFLAGS" }
             }
           }
           inverse_throughputs {
             source: "bhive: skl"
             inverse_throughput_cycles: 2.07
           })pb";
  EXPECT_THAT(x86_bhive_importer_->ParseBHiveCsvLine(
                  kSourceName,
                  "4829d38b44246c8b54246848c1fb034829d04839c3,207.000000", 0, 1,
                  /*throughput_scaling=*/kScaling, /*base_address=*/100),
              IsOkAndHolds(EqualsProto(kExpectedBasicBlockProto)));
}

TEST_F(BHiveImporterTest, InvalidColumnIndices) {
  EXPECT_THAT(x86_bhive_importer_->ParseBHiveCsvLine(kSourceName, "4929d2", 0,
                                                     1, kScaling),
              StatusIs(absl::StatusCode::kInvalidArgument));
}

TEST_F(BHiveImporterTest, InvalidColumnIndicesBig) {
  EXPECT_THAT(x86_bhive_importer_->ParseBHiveCsvLine(kSourceName,
                                                     "4929d2,a,b,c,5", 0, 10),
              StatusIs(absl::StatusCode::kInvalidArgument));
}

TEST_F(BHiveImporterTest, InvalidColumnIndicesSame) {
  EXPECT_THAT(
      x86_bhive_importer_->ParseBHiveCsvLine(kSourceName, "4929d2,5", 0, 0),
      StatusIs(absl::StatusCode::kInvalidArgument));
}

TEST_F(BHiveImporterTest, InvalidMachineCode) {
  // The binary code below is missing one byte at the end.
  EXPECT_THAT(x86_bhive_importer_->ParseBHiveCsvLine(kSourceName, "4929", 0, 1,
                                                     kScaling),
              StatusIs(absl::StatusCode::kInvalidArgument));
  EXPECT_THAT(x86_bhive_importer_->BasicBlockProtoFromMachineCodeHex("4929"),
              StatusIs(absl::StatusCode::kInternal));
}

TEST_F(BHiveImporterTest, NonStandardColumns) {
  EXPECT_THAT(x86_bhive_importer_->ParseBHiveCsvLine(kSourceName, "a,b,,0", 2,
                                                     3, kScaling),
              IsOkAndHolds(EqualsProto(R"pb(basic_block {}
                                            inverse_throughputs {
                                              source: "bhive: skl"
                                              inverse_throughput_cycles: 0
                                            })pb")));
}

TEST_F(BHiveImporterTest, MIRDatasetBasicTest) {
  EXPECT_THAT(x86_bhive_importer_->LoadMIRModule("sample_dataset/data.mir"),
              IsOk());
  EXPECT_THAT(x86_bhive_importer_->ParseMIRCsvLine(kSourceName, "a,b,BB_13,2.37", 2,
                                                     3, kScaling),
              IsOk());
}

}  // namespace
}  // namespace gematria
